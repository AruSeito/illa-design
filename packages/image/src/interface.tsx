import { SVGAttributes } from "react"

export interface ImageProps extends SVGAttributes<SVGElement> {
}