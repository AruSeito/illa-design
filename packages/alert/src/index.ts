export * from "./alert"
export * from "./interface"
