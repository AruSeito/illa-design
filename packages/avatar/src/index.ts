export * from "./avatar"
export * from "./interface"
export * from "./avatar-group"