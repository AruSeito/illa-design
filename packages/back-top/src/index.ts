export * from "./backtop"
export * from "./interface"
export * from "./style"
