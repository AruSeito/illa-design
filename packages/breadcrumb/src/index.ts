export * from "./interface"
export * from "./breadcrumb"
export * from "./breadcrumbItem"
