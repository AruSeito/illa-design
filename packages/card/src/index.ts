export * from "./card"
export * from "./interface"
export * from "./grid"
export * from "./meta"
