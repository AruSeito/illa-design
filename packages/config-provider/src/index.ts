export * from "./config-provider"
export * from "./interface"

// locale
export * from "./locale/en-US"
export * from "./locale/zh-CN"
export * from "./locale/def"
