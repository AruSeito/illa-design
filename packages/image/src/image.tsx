/** @jsxImportSource @emotion/react */
import * as React from "react"
import { forwardRef, useEffect, useState } from "react"
import { ImageProps } from "./interface"
import { ImageDefaultIcon } from "@illa-design/icon"
import { css } from "@emotion/core"
import { globalColor, illaPrefix } from "@illa-design/theme"

enum ImageState {
  Loading,
  Loaded,
  Error
}

function applyImageCss(objFit: "fill" | "container" | "cover" | "none" | "scale-down", radius: string) {
  return css`
    object-fit: ${objFit};
    background-color: ${globalColor(`--${illaPrefix}-blackAlpha-09`)};
    border-radius: ${radius};
  `
}

function applyDefaultFallback(width: string | number, height: string | number, radius: string) {
  return css`
    display: inline-flex;
    flex-direction: row;
    align-items: center;
    justify-content: center;
    background-color: ${globalColor(`--${illaPrefix}-blackAlpha-09`)};
    border-radius: ${radius};
    width: ${width};
    height: ${height};
  `
}

const applyOuterCss = css`
  display: inline-flex;
`

export const Image = forwardRef<HTMLDivElement, ImageProps>((props, ref) => {
  const {
    // origin
    src = "",
    width = "100px",
    height = "100px",
    alt = "",
    // custom
    objectFit = "cover",
    fallback = <ImageDefaultIcon />,
    fallbackSrc = "",
    radius = "4px",
    // img
    crossOrigin,
    decoding,
    loading,
    referrerPolicy,
    sizes,
    srcSet,
    useMap,
    ...rest
  } = props

  const [imageState, setImageState] = useState<ImageState>()

  useEffect(() => {
    setImageState(ImageState.Loading)
  }, [src])

  return <div css={applyOuterCss} ref={ref} {...rest}>
    {(src && src.length != 0) && imageState != ImageState.Error ?
      <img css={applyImageCss(objectFit, radius)}
           alt={alt}
           src={src}
           width={width}
           height={height}
           onError={
             (e) => {
               setImageState(ImageState.Error)
               if (props.onError != undefined) {
                 props.onError(e)
               }
             }
           }
           onLoad={
             (e) => {
               setImageState(ImageState.Loaded)
               if (props.onLoad != undefined) {
                 props.onLoad(e)
               }
             }
           }
           crossOrigin={crossOrigin}
           decoding={decoding}
           loading={loading}
           referrerPolicy={referrerPolicy}
           sizes={sizes}
           srcSet={srcSet}
           useMap={useMap}
      /> : fallbackSrc && fallbackSrc.length != 0 ?
        <img css={applyImageCss(objectFit, radius)}
             alt={alt}
             src={fallbackSrc}
             width={width}
             height={height}
             crossOrigin={crossOrigin}
             decoding={decoding}
             loading={loading}
             referrerPolicy={referrerPolicy}
             sizes={sizes}
             srcSet={srcSet}
             useMap={useMap}
        /> : <div css={applyDefaultFallback(width, height, radius)}>
          {fallback}
        </div>
    }
  </div>
})
