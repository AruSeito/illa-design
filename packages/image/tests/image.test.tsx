import { create } from "react-test-renderer"
import { Image } from "../src"
import { BsFacebook } from "react-icons/bs"
import { fireEvent, getByRole, render } from "@testing-library/react"

test("Image renders with default fallback", () => {
  const node = create(<Image className="test-image-render-fallback-default" />)
  expect(node.root.findByType(Image).props.className).toBe("test-image-render-fallback-default")
})

test("Image renders with fallback src", () => {
  const node = create(
    <Image className="test-image-render-fallback-src"
           fallbackSrc={"https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png"} />,
  )
  expect(node.root.findByType(Image).props.className).toBe("test-image-render-fallback-src")
})

test("Image renders with fallback icon", () => {
  const node = create(
    <Image className="test-image-render-fallback-icon"
           fallback={<BsFacebook />} />,
  )
  expect(node.root.findByType(Image).props.className).toBe("test-image-render-fallback-icon")
})

test("Image renders with size", () => {
  const node = create(
    <Image height="200px" width="200px"
           src="https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png" />,
  ).toJSON()
  expect(node).toMatchSnapshot()
})

test("Image renders with event", () => {
  const errorHandler = jest.fn()
  const loadHandler = jest.fn()
  const { getByRole } = render(
    <Image onError={errorHandler} onLoad={loadHandler}
           src="https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png" />,
  )
  fireEvent.load(getByRole("img"))
  expect(loadHandler).toBeCalled()
  fireEvent.error(getByRole("img"))
  expect(errorHandler).toBeCalled()
})