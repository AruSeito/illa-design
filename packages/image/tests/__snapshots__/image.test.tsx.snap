// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Image renders with size 1`] = `
.emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-1 {
  object-fit: cover;
  background-color: rgba(0, 0, 0, 0.04);
  border-radius: 4px;
}

<div
  className="emotion-0"
>
  <img
    alt=""
    className="emotion-1"
    height="200px"
    onError={[Function]}
    onLoad={[Function]}
    src="https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png"
    width="200px"
  />
</div>
`;
